notebooks/writing_a_trace_transform_cpu_offloading.ipynb

```diff
@@ -0,0 +1,868 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import gc\n",
+    "from typing import Callable, Mapping\n",
+    "\n",
+    "import torch\n",
+    "import torch.utils.benchmark\n",
+    "\n",
+    "import thunder\n",
+    "from thunder.core.trace import TraceCtx\n",
+    "from thunder.core.transform_common import Transform\n",
+    "from thunder.core.proxies import TensorProxy, variableify, Variable\n",
+    "from thunder.core.pytree import tree_map\n",
+    "from thunder.core.trace import tracectx, from_trace\n",
+    "from thunder.extend import OperatorExecutor\n",
+    "from thunder.core.symbol import BoundSymbol\n",
+    "from thunder.core import prims\n",
+    "from thunder.core.transforms import bsym_list_to_dag, Node, toposort_bsym_dag, TOPOSORT_ORDER\n",
+    "from thunder.core.vjp_utils import get_saved_for_backward_tensors\n",
+    "from thunder.core.module import ThunderModule"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Transforms\n",
+    "\n",
+    "To understand transforms, we need to know what a `Trace` is. In `thunder`, `Trace` is the representation of the jitted program in terms of thunder operations/symbol. Each operation in `Trace` is a collection of `BoundSymbol` i.e. a `Symbol` with it's input and output. We can also print the trace as a Python program for easier inspection. We will do this later in the notebook. To understand these concepts, you can read the helpful `zero_to_thunder.ipynb`. \n",
+    "\n",
+    "`thunder` allows us to write our custom transforms to transform trace/s. These transforms can be used for replace pointwise operations with fused implementation, compute gradient of the given computation, etc. Besides this, `thunder` enables us to apply these transforms at different stages during compilation. In this tutorial, we will use the post optimization stage, the point at which we already have the forward and the backward execution trace ready. To write our transform, we have to inherit from `Transform` class. This class implements the interface that each transform should have. By default, it provides no-op transformations. To use our transform, we provide an instance of our transform object to the `thunder.jit` via `transforms` argument.\n",
+    "\n",
+    "However, before writing our transform, we will make an `OperatorExecutor` with which we will create 2 operators/symbol - 1. to offload tensors to CPU 2. Load the offloaded tensors back to CUDA device. You read more about adding custom operators in `adding_custom_operator.ipynb` and also `zero_to_thunder.ipynb`."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Create a new executor.\n",
+    "offload_ex = OperatorExecutor(\"offload_ex\")\n",
+    "\n",
+    "\n",
```

Comment:
I see you ended up using pinned memory for the cpu offload. Did you get a chance to try streaming + pinned memory in conjunction at all? Also would it be worthwhile briefly adding some comment about why pinned memory is needed for cpu offloading and tradeoff vs streams only, just as background info?

Reply:
As of now, I haven't investigated more into `streaming + pinned memory`. 

> Also would it be worthwhile briefly adding some comment about why pinned memory is needed for cpu offloading and tradeoff vs streams only, just as background info?

I will add the link to [tutorial ](https://pytorch.org/tutorials/intermediate/pinmem_nonblock.html) regarding pinned_memory and non_blocking copies. I think let's refrain from `tradeoff vs stream only` as I am not very sure if that implementation was most optimal.